---
output:
    html_document:
        keep_md: TRUE
---

<script src="../assets/js/quiz.js"></script>

```{css echo=FALSE}
.colsel {
background-color: lightyellow;
}

pre, code {
  white-space:pre !important;
  overflow-x:scroll auto
}
```


```{r echo=FALSE}
library(knitr)
colFmt = function(x,color){
  outputFormat = opts_knit$get("rmarkdown.pandoc.to")
  if(outputFormat == 'latex')
    paste("\\textcolor{",color,"}{",x,"}",sep="")
  else if(outputFormat == 'html')
    paste("<font color='",color,"'>",x,"</font>",sep="")
  else
    x
}
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(class.source = "colsel", echo=TRUE)
```

```{r echo=F, include=FALSE}
library(dplyr)
library(kableExtra)
```


R and RStudio
========================================================

### What is R?
[R](http://r-project.org/) is a language and environment for statistical computing and graphics developed in 1993. It provides a wide variety of statistical and graphical techniques (linear and nonlinear modeling, statistical tests, time series analysis, classification, clustering, ...), and is highly extensible, meaning that the user community can write new R tools. It is a GNU project (Free and Open Source).

The R language has its roots in the S language and environment which was developed at Bell Laboratories (formerly AT&T, now Lucent Technologies) by John Chambers and colleagues. R was created by Ross Ihaka and Robert Gentleman at the University of Auckland, New Zealand, and now, R is developed by the R Development Core Team, of which Chambers is a member. R is named partly after the first names of the first two R authors (Robert Gentleman and Ross Ihaka), and partly as a play on the name of S. R can be considered as a different implementation of S. There are some important differences, but much code written for S runs unaltered under R.

Some of R's strengths:

* The ease with which well-designed publication-quality plots can be produced, including mathematical symbols and formulae where needed. Great care has been taken over the defaults for the minor design choices in graphics, but the user retains full control. Such as examples like the following (extracted from http://web.stanford.edu/class/bios221/book/Chap-Graphics.html):

```{r echo=FALSE, message=FALSE, error=FALSE, out.width="32%", fig.align='default'}
library(knitr)
include_graphics(c("chap3-rgraphics-heatmap-1.png", "chap3-rgraphics-darned1-1.png", "chap3-rgraphics-twodsp4-1.png"))
```

* It compiles and runs on a wide variety of UNIX platforms and similar systems (including FreeBSD and Linux), Windows and MacOS.
* R can be extended (easily) via packages.
* R has its own LaTeX-like documentation format, which is used to supply comprehensive documentation, both on-line in a number of formats and in hardcopy.
* It has a vast community both in academia and in business.
* It's FREE!   

### The R environment
R is an integrated suite of software facilities for data manipulation, calculation and graphical display. It includes

* an effective data handling and storage facility,
* a suite of operators for calculations on arrays, in particular matrices,
* a large, coherent, integrated collection of intermediate tools for data analysis,
* graphical facilities for data analysis and display either on-screen or on hardcopy, and
* a well-developed, and effective programming language which includes conditionals, loops, user-defined recursive functions and input and output facilities.

The term "environment" is intended to characterize it as a fully planned and coherent system, rather than an incremental accretion of very specific and inflexible tools, as is frequently the case with other data analysis software.

R, like S, is designed around a true computer language, and it allows users to add additional functionality by defining new functions. Much of the system is itself written in the R dialect of S, which makes it easy for users to follow the algorithmic choices made. For computationally-intensive tasks, C, C++ and Fortran code can be linked and called at run time. Advanced users can write C code to manipulate R objects directly.

Many users think of R as a statistics system. The R group prefers to think of it of an environment within which statistical techniques are implemented.   

### The R Homepage
The R homepage has a wealth of information on it,

[R-project.org](http://r-project.org/)

On the homepage you can:

* Learn more about R
* Download R
* Get Documentation (official and user supplied)
* Get access to CRAN 'Comprehensive R archival network'

### Interface for R

There are many ways one can interface with R language. Here are a few popular ones:

* [RStudio](https://raw.githubusercontent.com/ucdavis-bioinformatics-training/2019-March-Bioinformatics-Prerequisites/master/wednesday/Intro2R/rstudio.png)
* [RGui](https://raw.githubusercontent.com/ucdavis-bioinformatics-training/2019-March-Bioinformatics-Prerequisites/master/wednesday/Intro2R/rgui.png)
* Jupyter and R notebooks
* text editors, such as vi(m), Emacs...


### RStudio

[RStudio](http://rstudio.com/) started in 2010, to offer R a more full featured integrated development environment (IDE) and modeled after matlab's IDE.

RStudio has many features:

* syntax highlighting
* code completion
* smart indentation
* "Projects"
* workspace browser and data viewer
* embedded plots
* Markdown notebooks, Sweave authoring and knitr with one click pdf or html
* runs on all platforms and over the web
* etc. etc. etc.

RStudio and its team have contributed to many R packages.[13] These include:

* Tidyverse – R packages for data science, including ggplot2, dplyr, tidyr, and purrr
* Shiny – An interactive web technology
* RMarkdown – Insert R code into markdown documents
* knitr – Dynamic reports combining R, TeX, Markdown & HTML
* packrat – Package dependency tool
* devtools – Package development tool

RStudio Cheat Sheets: [rstudio-ide.pdf](https://github.com/rstudio/cheatsheets/raw/master/rstudio-ide.pdf)

---

Programming fundamentals
====================================================


There are three concepts that are essential in any programming language:

* `r colFmt("VARIABLES", 'purple')`

A variable is a named storage. Creating a variable is to reserve some space in memory. In R, the name of a variable can have letters, numbers, dot and underscore. However, a valid variable name cannot start with a underscore or a number, or start with a dot that is followed by a number.


* `r colFmt("FUNCTIONS", 'purple')`

A function is a block of organized, reusable code that is used to perform a set of predefined operations. A function may take zero or more parameters and return a result.

```{r out.width="15%", fig.align='center', echo=FALSE, message=F}
include_graphics("./func.png")
```


The way to use a function in R is:

**function.name(parameter1=value1, ...)**

In R, to get help information on a funciton, one may use the command: 

**?function.name**


* `r colFmt("OPERATIONS", 'purple')`

```{r echo=FALSE, results= 'asis'}
library(knitr)
library(kableExtra)
kable(data.frame(Operator=paste("<-", " =", sep=","), Description="Assignment", stringsAsFactors=F), align='c', caption="Assignment Operators in R") %>% kable_styling(bootstrap_options="striped", full_width=F, position="center")

kable(data.frame(Operator=c("+", "-", "*", "/", "^", "%%", "%/%"), Description=c("Addition", "Subtraction", "Multiplication", "Division", "Exponent", "Modulus", "Integer Division"), stringsAsFactors=F), align='c', caption="Arithmetic Operators in R") %>% kable_styling(bootstrap_options="striped", full_width=F, position="center")

kable(data.frame(Operator=c("<", ">", "<=", ">=", "==", "!="), Description=c("Less than", "Greater than", "Less than or equal to", "Greater than or equal to", "Equal to", "Not equal to"), stringsAsFactors=F), align='c', caption="Relational Operators in R") %>% kable_styling(bootstrap_options="striped", full_width=F, position="center")

kable(data.frame(Operator=c("!", "&", "&&", "|", "||"), Description=c("Logical NOT", "Element-wise logical AND", "Logical AND", "Element-wise logical OR", "Logical OR"), stringsAsFactors=F), align='c', caption="Logical Operators in R") %>% kable_styling(bootstrap_options="striped", full_width=F, position="center")
```

---


Start an R session
====================
**BEFORE YOU BEGIN, YOU NEED TO START AN R SESSION**

You can run this tutorial in an IDE (like Rstudio) on your laptop, or you can run R on the command-line on tadpole by logging into tadpole in a terminal and running the following commands:

> module load R
>
> R

**NOTE: Below, the text in the yellow boxes is code to input (by typing it or copy/pasting) into your R session, the text in the white boxes is the expected output.**

---

Topics covered in this introduction to R
====================================================
1. Basic data types in R
2. Basic data structures in R
3. Import and export data in R
4. Functions in R
5. Basic statistics in R
6. Simple data visulization in R
7. Install packages in R
8. Save data in R session
9. R markdown and R notebooks

---

Topic 1. Basic data types in R
====================================================

#### There are 5 basic atomic classes: numeric (integer, complex), character, logical


##### Examples of numeric values.

```{r}
# assign number 150 to variable a.
a <- 150
a
# assign a number in scientific format to variable b.
b <- 3e-2
b
```
<br>

##### Examples of character values.

```{r}
# assign a string "BRCA1" to variable gene
gene <- "BRCA1"
gene
# assign a string "Hello World" to variable hello
hello <- "Hello World"
hello
```
<br>

##### Examples of logical values.

```{r}
# assign logical value "TRUE" to variable brca1_expressed
brca1_expressed <- TRUE
brca1_expressed
# assign logical value "FALSE" to variable her2_expressed
her2_expressed <- FALSE
her2_expressed
# assign logical value to a variable by logical operation
her2_expression_level <- 0
her2_expressed <- her2_expression_level > 0
her2_expressed
```
<br>

##### To find out the type of variable.

```{r}
class(her2_expressed)
# To check whether the variable is a specific type
is.numeric(gene)
is.numeric(a)
is.character(gene)

```
<br>

##### In the case that one compares two different classes of data, the coersion rule in R is `r colFmt("logical -> integer -> numeric -> complex -> character", 'red')` . The following is an example of converting a numeric variable to character.


```{r}
b
as.character(b)
```
<br>

What happens when one converts a logical variable to numeric?

```{r}
# recall her2_expressed
her2_expressed

# conversion
as.numeric(her2_expressed)

her2_expressed + 1

```
<br>

##### A logical *TRUE* is converted to integer 1 and a logical *FALSE* is converted to integer 0.

<br>

## Quiz 1

<div id="quiz1" class="quiz"></div>
<button id="submit1">Submit Quiz</button>
<div id="results1" class="output"></div>
<script>
quizContainer1 = document.getElementById('quiz1');
resultsContainer1 = document.getElementById('results1');
submitButton1 = document.getElementById('submit1');

myQuestions1 = [
  {
    question: "Create a variable a and set it to 3, and a variable b set to 'gene'. What is a + b?",
    answers: {
      a: "a",
      b: "3",
      c: "Gives an error",
      d: "4"
    },
    correctAnswer: "c"
  },
  {
    question: "Create another variable c set to FALSE. What is a + c?",
    answers: {
      a: "Gives an error",
      b: "3",
      c: "a",
      d: "4"
    },
    correctAnswer: "b"
  },
  {
    question: "What is 1 + TRUE?",
    answers: {
      a: "2",
      b: "1",
      c: "TRUE",
      d: "FALSE"
    },
    correctAnswer: "a"
  }
];

buildQuiz(myQuestions1, quizContainer1);
submitButton1.addEventListener('click', function() {showResults(myQuestions1, quizContainer1, resultsContainer1);});
</script>

---


Topic 2. Basic data structures in R
====================================================

```{r echo=FALSE}

kable(data.frame(Homogeneous=c("Atomic vector", "Matrix", "Array"), Heterogeneous=c("List", "Data frame", ""), row.names=c("1d", "2d", "Nd"), stringsAsFactors=F), align='c') %>% kable_styling(bootstrap_options="striped", full_width=F, position="center", font_size=18)
```
<br>


#### Atomic vectors: an atomic vector is a combination of multiple values(numeric, character or logical) in the same object. An atomic vector is created using the function c().

```{r}
gene_names <- c("ESR1", "p53", "PI3K", "BRCA1", "EGFR")
gene_names

gene_expression <- c(0, 100, 50, 200, 80)
gene_expression
```
<br>

##### One can give names to the elements of an atomic vector.
```{r}
# assign names to a vector by specifying them
names(gene_expression) <- c("ESR1", "p53", "PI3K", "BRCA1", "EGFR")
gene_expression

# assign names to a vector using another vector
names(gene_expression) <- gene_names
gene_expression
```
<br>

##### Or One may create a vector with named elements from scratch.
```{r}
gene_expression <- c(ESR1=0, p53=100, PI3K=50, BRCA1=200, EGFR=80)
gene_expression
```
<br>

##### To find out the length of a vector:
```{r}
length(gene_expression)
```

   
##### `r colFmt("NOTE: a vector can only hold elements of the same type. If there are a mixture of data types, they will be coerced according to the coersion rule mentioned earlier in this documentation.", 'red')`  

<br>

#### Factors: a factor is a special vector. It stores categorical data, which are important in statistical modeling and can only take on a limited number of pre-defined values. The function factor() can be used to create a factor.

```{r}
disease_stage <- factor(c("Stage1", "Stage2", "Stage2", "Stage3", "Stage1", "Stage4"))
disease_stage
```
<br>

##### In R, categories of the data are stored as factor levels. The function levels() can be used to access the factor levels.

```{r}
levels(disease_stage)
```

##### A function to compactly display the internal structure of an R object is str(). Please use str() to display the internal structure of the object we just created *disease_stage*. It shows that _disease_stage_ is a factor with four levels: "Stage1", "Stage2", "Stage3", etc... The integer numbers after the colon shows that these levels are encoded under the hood by integer values: the first level is 1, the second level is 2, and so on. Basically, when _factor_ function is called, R first scan through the vector to determine how many different categories there are, then it converts the character vector to a vector of integer values, with each integer value labeled with a category.


```{r}
str(disease_stage)
```


##### By default, R infers the factor levels by ordering the unique elements in a factor alphanumerically. One may specifically define the factor levels at the creation of the factor.


```{r}
disease_stage <- factor(c("Stage1", "Stage2", "Stage2", "Stage3", "Stage1", "Stage4"), levels=c("Stage2", "Stage1", "Stage3", "Stage4"))
# The encoding for levels are different from above.
str(disease_stage)
```

If you want to know the number of individuals at each levels, there are two functions: _summary_ and _table_.

```{r}
summary(disease_stage)
```

```{r}
table(disease_stage)
```


## Quiz 2

<div id="quiz2" class="quiz"></div>
<button id="submit2">Submit Quiz</button>
<div id="results2" class="output"></div>
<script>
quizContainer2 = document.getElementById('quiz2');
resultsContainer2 = document.getElementById('results2');
submitButton2 = document.getElementById('submit2');

myQuestions2 = [
  {
    question: "Create a new factor with levels specified. What happens when the factor contains elements that are not included in the levels?",
    answers: {
      a: "A new level will be added to the factor",
      b: "A new element will be added to the factor that is an NA",
      c: "Nothing happens",
      d: "Gives a warning"
    },
    correctAnswer: "b"
  },
  {
    question: "You can type a '?' and then a function name to get help for that function. What does the 'relevel' function do?",
    answers: {
      a: "Sorts the factors",
      b: "Overwrites the factor levels",
      c: "Adds a new level to the factors",
      d: "Reorders the levels"
    },
    correctAnswer: "d"
  },
  {
    question: "What would the levels be for the following vector as a factor:<br>c('a','C','d','b',1,'!')",
    answers: {
      a: "a b C d 1 !",
      b: "! 1 a b d C",
      c: "1 a b C d !",
      d: "! 1 a b C d"
    },
    correctAnswer: "d"
  }
];

buildQuiz(myQuestions2, quizContainer2);
submitButton2.addEventListener('click', function() {showResults(myQuestions2, quizContainer2, resultsContainer2);});
</script>


<br>

---

#### Matrices: A matrix is like an Excel sheet containing multiple rows and columns. It is used to combine vectors of the same type.

```{r}
col1 <- c(1,3,8,9)
col2 <- c(2,18,27,10)
col3 <- c(8,37,267,19)

my_matrix <- cbind(col1, col2, col3)
my_matrix

```

##### One other way to create a matrix is to use *matrix()* function.

```{r}
nums <- c(col1, col2, col3)
nums
matrix(nums, ncol=2)
```

```{r}
rownames(my_matrix) <- c("row1", "row2", "row3", "row4")
my_matrix

t(my_matrix)
```

##### To find out the dimension of a matrix:

```{r}
ncol(my_matrix)
nrow(my_matrix)
dim(my_matrix)
```


##### Calculations with numeric matrices.

```{r}
my_matrix * 3
log10(my_matrix)
```

Total of each row.
```{r}
rowSums(my_matrix)
```

Total of each column.
```{r}
colSums(my_matrix)
```

##### There is a data structure _Array_ in R, that holds multi-dimensional (d > 2) data and is a generalized version of a matrix. *Matrix* is used much more commonly than *Array*, therefore we are not going to talk about *Array* here.


#### Data frames: a data frame is like a matrix but can have columns with different types (numeric, character, logical).

##### A data frame can be created using the function data.frame().

```{r}
# creating a data frame using pre-defined vectors
patients_name=c("Patient1", "Patient2", "Patient3", "Patient4", "Patient5", "Patient6")
Family_history=c("Y", "N", "Y", "N", "Y", "Y")
patients_age=c(31, 40, 39, 50, 45, 65)
meta.data <- data.frame(patients_name=patients_name, disease_stage=disease_stage, Family_history=Family_history, patients_age=patients_age)
meta.data
```

###### To check whether a data is a data frame, use the function is.data.frame().

```{r}
is.data.frame(meta.data)
```

```{r}
is.data.frame(my_matrix)
```

###### One can convert a matrix object to a data frame using the function as.data.frame().

```{r}
class(my_matrix)
```

```{r}
my_data <- as.data.frame(my_matrix)
class(my_data)
```

##### A data frame can be transposed in the similar way as a matrix. However, the result of transposing a data frame might not be a data frame anymore.

```{r}
my_data
```

```{r}
t(my_data)
```

##### A data frame can be extended.

```{r}
# add a column that has the information on harmful mutations in BRCA1/BRCA2 genes for each patient.
meta.data
meta.data$BRCA <- c("YES", "NO", "YES", "YES", "YES", "NO")
meta.data
```

##### A data frame can also be extended using the functions cbind() and rbind(), for adding columns and rows respectively. When using cbind(), the number of values in the new column must match the number of rows in the data frame. When using rbind(), the two data frames must have the same variables/columns.

```{r}
# add a column that has the information on the racial information for each patient.
cbind(meta.data, Race=c("AJ", "AS", "AA", "NE", "NE", "AS"))
# rbind can be used to add more rows to a data frame.
rbind(meta.data, data.frame(patients_name="Patient7", disease_stage="S4", Family_history="Y", patients_age=48, BRCA="YES"))
```

##### One may use the function *merge* to merge two data frames horizontally, based on one or more common key variables.

```{r}
expression.data <- data.frame(patients_name=c("Patient3", "Patient4", "Patient5", "Patient1", "Patient2", "Patient6"), EGFR=c(10, 472, 103784, 1782, 187, 18289), TP53=c(16493, 72, 8193, 1849, 173894, 1482))
expression.data
md2 <- merge(meta.data, expression.data, by="patients_name")
md2
```


<br>

## Quiz 3

<div id="quiz3" class="quiz"></div>
<button id="submit3">Submit Quiz</button>
<div id="results3" class="output"></div>
<script>
quizContainer3 = document.getElementById('quiz3');
resultsContainer3 = document.getElementById('results3');
submitButton3 = document.getElementById('submit3');

myQuestions3 = [
  {
    question: "Find a function to add up the EGFR column in md2. What is the total?",
    answers: {
      a: "124524",
      b: "124526",
      c: "124528",
      d: "124530"
    },
    correctAnswer: "a"
  },
  {
    question: "Multiply my_matrix by itself, sum each column, and then use the 'mean' function to find the mean:",
    answers: {
      a: "24799.33",
      b: "24797.33",
      c: "24798.33",
      d: "24796.33"
    },
    correctAnswer: "c"
  }
];

buildQuiz(myQuestions3, quizContainer3);
submitButton3.addEventListener('click', function() {showResults(myQuestions3, quizContainer3, resultsContainer3);});
</script>


### Lists
#### A list is an ordered collection of objects, which can be any type of R objects (vectors, matrices, data frames, even lists).

##### A list is constructed using the function list().

```{r}
my_list <- list(1:5, "a", c(TRUE, FALSE, FALSE), c(3.2, 103.0, 82.3))
my_list
str(my_list)
```

##### One could construct a list by giving names to elements.

```{r}
my_list <- list(Ranking=1:5, ID="a", Test=c(TRUE, FALSE, FALSE), Score=c(3.2, 103.0, 82.3))

# display the names of elements in the list using the function *names*, or *str*. Compare the output of *str* with the above results to see the difference.
names(my_list)
str(my_list)
```

```{r}
# number of elements in the list
length(my_list)
```

### Subsetting data

#### Subsetting allows one to access the piece of data of interest. When combinded with assignment, subsetting can modify selected pieces of data. The operators that can be used to subset data are: [, $, and [[.

##### First, we are going to talk about subsetting data using [, which is the most commonly used operator. We will start by looking at vectors and talk about four ways to subset a vector.

* `r colFmt("**Positive integers** return elements at the specified positions", 'purple')`

```{r}
# first to recall what are stored in gene_names
gene_names
# obtain the first and the third elements
gene_names[c(1,3)]
```

R uses 1 based indexing, meaning the first element is at the position 1, not at position 0.

* `r colFmt("**Negative integers** omit elements at the specified positions", 'purple')`

```{r}
gene_names[-c(1,3)]
```

One may not mixed positive and negative integers in one single subset operation.

```{r error=TRUE}
# The following command will produce an error.
gene_names[c(-1, 2)]
```

* `r colFmt("**Logical vectors** select elements where the corresponding logical value is TRUE", 'purple')`, This is very useful because one may write the expression that creates the logical vector.

```{r}
gene_names[c(TRUE, FALSE, TRUE, FALSE, FALSE)]
```

Recall that we have created one vector called *gene_expression*. Let's assume that *gene_expression* stores the expression values correspond to the genes in *gene_names*. Then we may subset the genes based on expression values.

```{r}
gene_expression

gene_names[gene_expression > 50]
```

If the logical vector is shorter in length than the data vector that we want to subset, then it will be recycled to be the same length as the data vector.

```{r}
gene_names[c(TRUE, FALSE)]
```

If the logical vector has "NA" in it, the corresponding value will be "NA" in the output. "NA" in R is a symbol for missing value.

```{r}
gene_names[c(TRUE, NA, FALSE, TRUE, NA)]
```

* `r colFmt("**Character vectors** return elements with matching names, when the vector is named.", 'purple')`

```{r}
gene_expression
gene_expression[c("ESR1", "p53")]
```

* `r colFmt("**Nothing** returns the original vector", 'purple')`, This is more useful for matrices, data frames than for vectors.

```{r}
gene_names[]
```

<br>

##### Subsetting a list works in the same way as subsetting an atomic vector. Using [ will always return a list.

```{r}
my_list[1]
```
<br>

##### Subsetting a matrix can be done by simply generalizing the one dimension subsetting: one may supply a one dimension index for each dimension of the matrix. `r colFmt("Blank/Nothing subsetting is now useful in keeping all rows or all columns.", 'red')`


```{r}
my_matrix[c(TRUE, FALSE), ]
```
<br>

##### Subsetting a data frame can be done similarly as subsetting a matrix. In addition, one may supply only one 1-dimensional index to subset a data frame. In this case, R will treat the data frame as a list with each column is an element in the list.

```{r}
# recall a data frame created from above: *meta.data*
meta.data
# subset the data frame similarly to a matrix
meta.data[c(TRUE, FALSE, FALSE, TRUE),]

# subset the data frame using one vector
meta.data[c("patients_age", "disease_stage")]
```

<br>

### Subsetting operators: **[[** and **$**

##### **[[** is similar to **[**, except that it returns the content of the element.

```{r}
# recall my_list
my_list
# comparing [[ with [ in subsetting a list
my_list[[1]]
my_list[1]
```

`r colFmt("[[ is very useful when working with a list. Because when [ is applied to a list, it always returns a list. While [[ returns the contents of the list. [[ can only extrac/return one element, so it only accept one integer/string as input.", 'red')`

Because data frames are implemented as lists of columns, one may use [[ to extract a column from data frames.

```{r}
meta.data[["disease_stage"]]
```


<br>

##### **$** is a shorthand for **[[** combined with character subsetting.

```{r}
# subsetting a list using $ 
my_list$Score
# subsetting a data frame using
meta.data$disease_stage
```

<br>

##### Simplifying vs. preserving subsetting

We have seen some examples of simplying vs. preserving subsetting, for example:

```{r}
# simplifying subsetting
my_list[[1]]
# preserving subsetting
my_list[1]
```

Basically, simplying subsetting returns the simplest possible data structure that can represent the output. While preserving subsetting keeps the structure of the output as the same as the input. In the above example, [[ simplifies the output to a vector, while [ keeps the output as a list.

Because the syntax of carrying out simplifying and preserving subsetting differs depending on the data structure, the table below provides the information for the most basic data structure.

```{r, echo=FALSE}
kable(data.frame(Simplifying=c("x[[1]]", "x[[1]]", "x[1:3, drop=T]", "x[, 1] or x[[1]]"), Preserving=c("x[1]", "x[1]", "x[1:3]", "x[, 1, drop=F] or x[1]"), row.names=c("Vector", "List", "Factor", "Data frame"), stringsAsFactors=F), align='c') %>% kable_styling(bootstrap_options="striped", full_width=F, position="left")
```


---

Topic 3. Import and export data in R
====================================================

##### R base function read.table() is a general funciton that can be used to read a file in table format. The data will be imported as a data frame.

```{r}
# There is a very convenient way to read files from the internet.
data1 <- read.table(file="https://github.com/ucdavis-bioinformatics-training/courses/raw/master/Intro2R/raw_counts.txt", sep="\t", header=T, stringsAsFactors=F)

# To read a local file. If you have downloaded the raw_counts.txt file to your local machine, you may use the following command to read it in, by providing the full path for the file location. The way to specify the full path is the same as taught in the command line session.
download.file("https://github.com/ucdavis-bioinformatics-training/courses/raw/master/Intro2R/raw_counts.txt", "./raw_counts.txt")
data1 <- read.table(file="./raw_counts.txt", sep="\t", header=T, stringsAsFactors=F)

```

To check what type of object *data1* is in and take a look at the beginning part of the data.
```{r}
is.data.frame(data1)
head(data1)
```


##### Depending on the format of the file, several variants of read.table() are available to make reading a file easier.

* read.csv(): for reading "comma separated value" files (.csv).

* read.csv2(): variant used in countries that use a comma "," as decimal point and a semicolon ";" as field separators.

* read.delim(): for reading "tab separated value" files (".txt"). By default, point(".") is used as decimal point.

* read.delim2(): for reading "tab separated value" files (".txt"). By default, comma (",") is used as decimal point.

<br>

```{r}
# We are going to read a file over the internet by providing the url of the file.
data2 <- read.csv(file="https://github.com/ucdavis-bioinformatics-training/courses/raw/master/Intro2R/raw_counts.csv", stringsAsFactors=F)

# To look at the file:
head(data2)
```

<br>

##### R base function write.table() can be used to export data to a file.

```{r}
# To write to a file called "output.txt" in your current working directory.
write.table(data2[1:20,], file="output.txt", sep="\t", quote=F, row.names=T, col.names=T)
```

It is also possible to export data to a csv file.

write.csv()

write.csv2()


## Quiz 4

<div id="quiz4" class="quiz"></div>
<button id="submit4">Submit Quiz</button>
<div id="results4" class="output"></div>
<script>
quizContainer4 = document.getElementById('quiz4');
resultsContainer4 = document.getElementById('results4');
submitButton4 = document.getElementById('submit4');

myQuestions4 = [
  {
    question: "Using my_list, multiply the Ranking by the Score and find the mean in one command. What is the output?",
    answers: {
      a: "196.78 without a warning",
      b: "196.78 with a warning",
      c: "210.54 without a warning",
      d: "210.54 with a warning"
    },
    correctAnswer: "b"
  },
  {
    question: "Which of the following code will NOT get a result?",
    answers: {
      a: "gene_expression$ESR1",
      b: "gene_expression['ESR1']",
      c: "gene_expression[c(1,2,3,4)]",
      d: "gene_expression[1:2]"
    },
    correctAnswer: "a"
  },
  {
    question: "When you run this code:<br><br>my_list[1] * my_list[[1]]<br><br>you get an error. Why?",
    answers: {
      a: "Because you can't multiply a list and a number together.",
      b: "Because you can't multiply a list with itself.",
      c: "Because you can't multiply a list with another list.",
      d: "All of the above."
    },
    correctAnswer: "d"
  },
  {
    question: "Using data1 and the 'max' function, find the maximum value across columns C92, I563, and I861:",
    answers: {
      a: "69853",
      b: "112754",
      c: "88122",
      d: "66890"
    },
    correctAnswer: "a"
  }
];

buildQuiz(myQuestions4, quizContainer4);
submitButton4.addEventListener('click', function() {showResults(myQuestions4, quizContainer4, resultsContainer4);});
</script>


---

Topic 5. Functions in R
====================================================
#### Invoking a function by its name, followed by the parenthesis and zero or more arguments.

```{r}
# to find out the current working directory
getwd()

# to set a different working directory, use setwd
#setwd("/Users/jli/Desktop")

# to list all objects in the environment
ls()

# to create a vector from 2 to 3, using increment of 0.1
seq(2, 3, by=0.1)

# to create a vector with repeated elements
rep(1:3, times=3)
rep(1:3, each=3)

# to get help information on a function in R: ?function.name
?seq
?sort
?rep

```  

